Refactor and move wal docs (#16387)
hashicorp/consul: website/content/docs/agent/wal-logstore/monitoring.mdx

---
layout: docs
page_title: Monitor Raft metrics and logs for WAL
description: >-
  Consul 1.15 introduced a new experimental storage backend option. Learn how to

>   configure and test it out in a safe way.

Reviewer: ```suggestion
  Consul 1.15 introduced a new experimental storage backend option. Learn how to
  monitor performance and correctness while evaluating it.
```

---

# Monitor Raft metrics and logs for WAL

This topic describes how to monitor Raft metrics and logs if you are testing the WAL backend. We strongly recommend monitoring the Consul cluster, especially the target server, for evidence that the WAL backend is not functioning correctly. Refer to [Enable the experimental WAL LogStore backend](/consul/docs/agent/wal-logstore/index) for additional information about the WAL backend.

!> **Upgrade warning:** The WAL LogStore backend is experimental.

## Monitor for checksum failures

> Log store verification failures on any server, regardless of whether you are running the BoltDB or WAL backed, are unrecoverable error. Consul may report the following errors.

Reviewer: ```suggestion
Log store verification failures on any server, regardless of whether you are running the BoltDB or WAL backed, are unrecoverable errors. Consul may report the following errors in logs.
```

### Read failures: Disk Corruption

```log hideClipboard
2022-11-15T22:41:23.546Z [ERROR]  agent.raft.logstore: verification checksum FAILED: storage corruption rangeStart=1234 rangeEnd=3456 leaderChecksum=0xc1... readChecksum=0x45...
```

> This indicates that the server read back data that is different from what it wrote to disk, which signals corruption in the storage backend or filesystem.

Reviewer: ```suggestion
This indicates that the server read back data that is different from what it wrote to disk. This indicates corruption in the storage backend or filesystem.
```

For convenience, we also increment a metric `consul.raft.logstore.verifier.read_checksum_failures` when this occurs.

### Write failures: In-flight Corruption

> The following error indicates that the checksum on the follower did not match the leader when the leader _wrote_ the logs. The error implies that the corruption happened in the network or software and not the log store:

Reviewer: ```suggestion
The following error indicates that the checksum on the follower did not match the leader when the follower received the logs. The error implies that the corruption happened in the network or software and not the log store:
```

```log hideClipboard
2022-11-15T22:41:23.546Z [ERROR]  agent.raft.logstore: verification checksum FAILED: in-flight corruption rangeStart=1234 rangeEnd=3456 leaderChecksum=0xc1... followerWriteChecksum=0x45...
```

> It is unlikely that this error indicates an issue with the storage backend, but you should take steps to resolve it.

Reviewer: ```suggestion
It is unlikely that this error indicates an issue with the storage backend, but you should take the same steps to resolve and report it.
```

The `consul.raft.logstore.verifier.write_checksum_failures` increments when this error occurs.

## Resolve checksum failures

> If either type of corruption is detected, complete the instruction for [reverting to BoltDB](/consul/docs/agent/wal-logstore/revert-to-boltdb). If the server already uses BoltDB, the errors likely idicate a latent bug in BoltDB or a bug in the verification code, but you should still follow the revert instructions. 

Reviewer: ```suggestion
If either type of corruption is detected, complete the instructions for [reverting to BoltDB](/consul/docs/agent/wal-logstore/revert-to-boltdb). If the server already uses BoltDB, the errors likely indicate a latent bug in BoltDB or a bug in the verification code, but you should still follow the revert instructions. 
```
